### PR TITLE
[PyTorch] Add context parallel support for packed dataset in THD format

### PR DESCRIPTION
**Pull request by tomlifu**
# What does this PR do ?

This PR adds context parallel support for packed dataset in THD format in NeMo in response to this TE PR: https://github.com/NVIDIA/TransformerEngine/pull/641. Currently, the TE PR requires each individual sequence length is divisible by (2*context_parallel_size).

# Changes

- Add support to split packed dataset across different CP ranks in a load balanced way
- Add necessary paddings to dataset during packing stage to make sure the individual sequence length is a multiple of 2*cp_size

**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 1]
```diff
@@ -83,12 +83,26 @@ def tokenize_dataset(cfg: 'DictConfig'):
...
+    # if context parallel is used, each individual data length in one packed dataset sample
+    # needs to be a multiple of (cp_size * 2): https://github.com/NVIDIA/TransformerEngine/pull/641
+    if cp_size > 1:
+      # Function to calculate the LCM of two numbers
+      def lcm(x, y):
+        larger = max(x, y)
+        for i in range(larger, x * y + 1):
+          if i % x == 0 and i % y == 0:
+              return i
+      pad_seq_length_to_mult = lcm(pad_seq_length_to_mult, cp_size * 2)
```
Comment (xrennvidia) [inline scripts/nlp_language_modeling/prepare_packed_ft_dataset.py:98]: Do we need such a complex function? I guess `max(pad_seq_length_to_mult, cp_size*2)` is enough?

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 2]
```diff
@@ -109,8 +123,26 @@ def tokenize_dataset(cfg: 'DictConfig'):
...
+        '''
+        for key,val in data.items():
+          if key in {'input_ids', 'context_ids'}:
+            val = val + [pad_id] * (max_length - len(val) + 1) # add 1 for cp
+            data[key]=val
+        return
+      ceil_to_nearest = lambda n, m : (n + m - 1) // m * m
+      for data in dataset:
+        max_length = min(max_seq_length, ceil_to_nearest(len(data['input_ids']), pad_seq_length_to_mult))
+        assert max_length <= max_seq_length
```
Comment (xrennvidia) [inline scripts/nlp_language_modeling/prepare_packed_ft_dataset.py:143]: You have `min` in 142, so this assert seems redundant?

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 3]
```diff
@@ -109,8 +123,26 @@ def tokenize_dataset(cfg: 'DictConfig'):
...
+    pad_seq_length_to_mult = dataset.pad_seq_length_to_mult
+    dataset = np.array([dataset[i] for i in range(len(dataset))])
+    if cp_size > 1:
+      def pre_pad_dataset(data, max_length, pad_id):
+        '''
+        pad each individual data point to the length of max_length
+        '''
+        for key,val in data.items():
+          if key in {'input_ids', 'context_ids'}:
+            val = val + [pad_id] * (max_length - len(val) + 1) # add 1 for cp
```
Comment (xrennvidia) [inline scripts/nlp_language_modeling/prepare_packed_ft_dataset.py:137]: Why add 1 for CP?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 4]
```diff
@@ -1239,6 +1241,23 @@ def fwd_output_and_loss_func(dataloader_iter, model, checkpoint_activations_all_
...
 
+                    # get packed sequences for this context parallel rank
+                    cp_size = parallel_state.get_context_parallel_world_size()
+                    if cp_size > 1:
+                        cp_rank = parallel_state.get_context_parallel_rank()
+                        for key in required_keys:
+                            val = batch[key]
+                            if key not in {"cu_seqlens"}:
+                                index = tex.thd_get_partitioned_indices(cu_seqlens, val.size(1), cp_size, cp_rank)
+                                val = val.index_select(1, index)
```
Comment (xrennvidia) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1252]: This cannot handle attention_mask, right? Attention mask shape is [b, h, s, s,].

**Comment by xrennvidia** [comment]
You code indentation is not consistent, some places have 4 spaces, and other places have 2 spaces.
NeMo code has 4 spaces always, so please make sure all your code have 4-space indentation.

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py (hunk 1 above)
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py]: changed to `max(pad_seq_length_to_mult, cp_size*2)` in the latest commit

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py (hunk 2 above)
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py]: Removed the assert in the new commit

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 4 above)
Comment (tomlifu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py]: Added support for attention_mask in the new commit

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py (hunk 3 above)
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py]: I've documented the reason as a comment in the code: because input_ids is truncated by 1 for labels in the collate_fn of GPTSFTPackedDataset in gpt_sft_dataset.py, we add 1 extra padding here

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 5]
```diff
@@ -109,8 +117,29 @@ def tokenize_dataset(cfg: 'DictConfig'):
...
+                    # because input_ids is truncated by 1 for labels in the collate_fn of GPTSFTPackedDataset
+                    # in gpt_sft_dataset.py, we add 1 extra padding here
+                    val = val + [pad_id] * (max_length - len(val) + 1)
+                    data[key] = val
+            return
+
+        ceil_to_nearest = lambda n, m: (n + m - 1) // m * m
+        for data in dataset:
+            max_length = min(max_seq_length, ceil_to_nearest(len(data['input_ids']), pad_seq_length_to_mult))
+            pre_pad_dataset(data, max_length, pad_id)
```
Comment (xrennvidia) [inline scripts/nlp_language_modeling/prepare_packed_ft_dataset.py:141]: How the loss_mask is handled for padded tokens?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 6]
```diff
@@ -1261,6 +1263,26 @@ def fwd_output_and_loss_func(dataloader_iter, model, checkpoint_activations_all_
...
+                                seq_dim = 1 if key != 'attention_mask' else 2
+                                index = tex.thd_get_partitioned_indices(
+                                    cu_seqlens, val.size(seq_dim), cp_size, cp_rank
+                                )
+                                val = val.index_select(seq_dim, index)
+                                batch[key] = val
+                        cu_seqlens = cu_seqlens // cp_size
+                        forward_args = {
+                            'input_ids': batch['tokens'],
+                            'position_ids': batch['position_ids'],
```
Comment (xrennvidia) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1282]: the position_ids means the token_id in packed sequence? how is this argument used in training fwd and bwd?

**Comment by xrennvidia** [comment]
Thanks for fixing the comments, it looks much better now.
The total sequence length (size of t in THD format) is a constant, right? If so, we should have some padded tokens at the end? how those padded tokens are split across different CP ranks?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 6 above)
Comment (tomlifu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1282]: The position_ids is the position of the tokens in a sequence (e.g. [0,1,2, ... , seq_len-1]). In a packed sequence, we have a list of position_ids since the packed sequence is composed of many individual sequences.  I'm not too sure if that's what you mean by token_id. It's used the same way as input_ids in training fwd and bwd.

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py (hunk 5 above)
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py:141]: The loss_mask is handled in the packing function here: https://github.com/NVIDIA/NeMo/blob/main/scripts/nlp_language_modeling/prepare_packed_ft_dataset.py#L142

**Comment by youth123** [comment]
> # What does this PR do ?
> This PR adds context parallel support for packed dataset in THD format in NeMo in response to this TE PR: [NVIDIA/TransformerEngine#641](https://github.com/NVIDIA/TransformerEngine/pull/641). Currently, the TE PR requires each individual sequence length is divisible by (2*context_parallel_size).
> 
> # Changes
> * Add support to split packed dataset across different CP ranks in a load balanced way
> * Add necessary paddings to dataset during packing stage to make sure the individual sequence length is a multiple of 2*cp_size
> 
> **PR Type**:
> 
> * [x]  New Feature
> * [ ]  Bugfix
> * [ ]  Documentation
Hello, I have some problems with cp need seqlen divisibility world_size * 2. I see that you are padding data in the code, but this will cause the pad token id to enter the attention calculation. I don't know if this is correct.
